johnnydep/lib.py

```diff
@@ -39,14 +40,38 @@ def __missing__(self, key):
         return value
 
 
-class JohnnyDist(anytree.NodeMixin):
-    def __init__(self, req_string, parent=None, index_url=None, env=None, extra_index_url=None):
+class JohnnyBase(anytree.NodeMixin):
+    def __init__(self, parent=None):
+        self.parent = parent
+
+
+class JohnnyDistFailed(JohnnyBase):
+    def __init__(self, req_string, parent=None, error=None):
+        self.req = pkg_resources.Requirement.parse(req_string)
+        self.req.name = "(Failed) " + self.req.name
+        self.name = canonicalize_name(self.req.name)
+        self.specifier = str(self.req.specifier)
+        try:
+            output = error.decode("utf-8")
+            lines = output.split('\n')
+            self.error = lines[-3]
+        except (ValueError, AttributeError):
+            self.error = "Exception: Unable to parse exception message"
+        super(JohnnyDistFailed, self).__init__(parent)
```

Comment:
This class is missing many of the attributes that a `JohnnyDist` instance needs to have (stuff that normally would be set in `JohnnyDist` init).  For example when triggering a failure with `johnnydep wandb --ignore-errors -o yaml` I get:

```
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\wglenn\git\johnnydep\johnnydep\__main__.py", line 4, in <module>
    main()
  File "C:\Users\wglenn\git\johnnydep\johnnydep\cli.py", line 85, in main
    print(dist.serialise(fields=args.fields, format=args.output_format, recurse=args.recurse))
  File "C:\Users\wglenn\git\johnnydep\johnnydep\lib.py", line 282, in serialise
    next(deps)  # skip over root
  File "C:\Users\wglenn\git\johnnydep\johnnydep\lib.py", line 339, in flatten_deps
    extra_map[dep.name] |= set(dep.extras_requested)
AttributeError: 'JohnnyDistFailed' object has no attribute 'extras_requested'
```

I think you should try to implement this feature by handling errors within `JohnnyDist` directly, without splitting into two separate types.

Reply:
Thanks for reviewing. I have no issue re-writing this but I'm curious why you think it would be better to handle the errors within JohnnyDist instead of with inheritance? My initial thought was that one might want to distinguish between a successful and a failed JonnyDist type.

Reply:
They can be distinguished with an attribute.  They're not so different enough that distinct types are necessary.